### ISSUE TITLE
AF fails on file

### ISSUE THREAD

**Issue by LandonTClipp** [TerraFusion/advancedFusion#11]
The AF tool encounters an HDF error when using this configuration:
/u/sciteam/clipp/scratch/af_out/logs/run22/configs/37610_config.txt
```
MISR_RADIANCE: Blue_Radiance Green_Radiance
OUTPUT_FILE_PATH: /u/sciteam/clipp/scratch/af_out/data/2007.01/ADVNCE_FUSE_TERRA_BF_L1B_O37610_20070112235819_F000_V001.h5
SOURCE_INSTRUMENT: MISR
MISR_RESOLUTION: L
MODIS_BANDS: 8 9
MODIS_RESOLUTION: 1KM
RESAMPLE_METHOD: nnInterpolate
INPUT_FILE_PATH: /u/sciteam/clipp/scratch/basicfusion/2007.01/TERRA_BF_L1B_O37610_20070112235819_F000_V001.h5
MISR_CAMERA_ANGLE: AN AF
TARGET_INSTRUMENT: MODIS
```

The HDF error is:

```
User input handling ...

Getting source instrument latitude & longitude data...
Retrieveing latitude data for MISR
HDF5-DIAG: Error detected in HDF5 (1.10.1) thread 0:
  #000: H5D.c line 294 in H5Dopen2(): unable to open dataset
    major: Dataset
    minor: Can't open object
  #001: H5Dint.c line 1362 in H5D__open_name(): not found
    major: Dataset
    minor: Object not found
  #002: H5Gloc.c line 428 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #003: H5Gtraverse.c line 867 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #004: H5Gtraverse.c line 753 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
Dataset open error
HDF5-DIAG: Error detected in HDF5 (1.10.1) thread 0:
  #000: H5D.c line 294 in H5Dopen2(): unable to open dataset
    major: Dataset
    minor: Can't open object
  #001: H5Dint.c line 1362 in H5D__open_name(): not found
    major: Dataset
    minor: Object not found
  #002: H5Gloc.c line 428 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #003: H5Gtraverse.c line 867 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #004: H5Gtraverse.c line 753 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
Dataset open error
Segmentation fault
```

**Comment by kyang2014** [comment]
You should first check if the BF file is corrupt or missing the latitude and longitude fields. I am also curious why HDF5 1.10.1 is used. We are supposed to use 1.8.??

**Comment by LandonTClipp** [comment]
I checked with h5dump and it doesn't report the file as corrupt. I actually have 1.8, forgot that advanced fusion uses 1.8 too. 

@jkimdevnc Do you have any ideas?

**Comment by jkimdevnc** [comment]
The error output indicates that it’s occurring while AFtool is opening a MISR’s latitude dataset from the BF file.
Basically, a latitude dataset is missing from the BF file.

Unfortunately, I am not in a situation to be able to work on this directly until next week.

Could you try with a different BF file?
If the same error occurs, it can be AFtool issue and need to be debugged to figure out what’s going on.
However, if the error doesn’t occur with other BF file(s), it’s a data problem as the BF file is missing a latitude dataset.
(ex: no issue with “TERRA_BF_L1B_O69400_20130104000439_F000_V000.h5”)

If you use BW, make sure you run “$ module load cray-hdf5/1.8.16” before running AFtool.

Jonathan

**Comment by kyang2014** [comment]
Just check the CDL and the input file list of 37610. This file doesn't have any MISR data. The AF tool seems not considering this case and generate an error.

ALso note:

Landon's version of BF is TERRA_BF_L1B_O37610_20070112235819_F000_V001.h5

Jonthan's version of BF is TERRA_BF_L1B_O69400_20130104000439_F000_V000.h5

If time allows, we should use the V001.h5 to test.
On Jun 6, 2018, at 6:33 PM, jkimdevnc <notifications@github.com<mailto:notifications@github.com>> wrote:

The error output indicates that it’s occurring while AFtool is opening a MISR’s latitude dataset from the BF file.
Basically, a latitude dataset is missing from the BF file.

Unfortunately, I am not in a situation to be able to work on this directly until next week.

Could you try with a different BF file?
If the same error occurs, it can be AFtool issue and need to be debugged to figure out what’s going on.
However, if the error doesn’t occur with other BF file(s), it’s a data problem as the BF file is missing a latitude dataset.
(ex: no issue with “TERRA_BF_L1B_O69400_20130104000439_F000_V000.h5”)

If you use BW, make sure you run “$ module load cray-hdf5/1.8.16” before running AFtool.

Jonathan

**Comment by jkimdevnc** [comment]
Yes, if so, a tool should be updated to handle such a case more gracefully.
Landon, sure please go ahead.

**Comment by LandonTClipp** [comment]
AF tool now handles missing dataset for MISR a little more gracefully, however there are still unclosed file IDs at program termination. I have also begun to implement better stack tracing in the program... for example when running on the configuration file given:

```
Getting source instrument latitude & longitude data...
Retrieveing latitude data for MISR
HDF5-DIAG: Error detected in HDF5 (1.8.16) thread 0:
  #000: H5D.c line 358 in H5Dopen2(): not found
    major: Dataset
    minor: Object not found
  #001: H5Gloc.c line 430 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #002: H5Gtraverse.c line 861 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #003: H5Gtraverse.c line 755 in H5G_traverse_real(): component not found
    major: Symbol table
    minor: Object not found
[io.cpp:2287:af_read()] ERROR:
        On call: dataset = H5Dopen2(file, dataset_name, H5P_DEFAULT)
        returned: -1
[io.cpp:188:get_misr_lat()] ERROR:
        On call: lat_data = af_read(file, lat_dataset_name)
        returned: 0
[AFtool.cpp:81:AF_GetGeolocationDataFromInstrument()] ERROR:
        On call: *latitude = get_misr_lat(inputFile, (char*) resolution.c_str(), &cellNum)
        returned: 0
[AFtool.cpp:559:main()] ERROR:
        On call: ret = AF_GetGeolocationDataFromInstrument(srcInstrument, inputArgs, inputFile, &srcLatitude , &srcLongitude , srcCellNum )
        returned: -1
```

I will be going through to clean up a bit more.

93eb173aea2bcd5cedb8ab1338429fb67930b26c

**Comment by jkimdevnc** [comment]
Thanks.
FYI, there is “AF_debug.h” we can use to set for debugging information.    You can integrate the tracing output to one of those as well.

**Comment by LandonTClipp** [comment]
I just ran two months of AF using my parallel workflow. It took 10 minutes. I am attaching a list of trouble orbits that the AF tool cannot handle. Most of these are probably missing datasets. We should discuss the behavior of AF in this case... should it ouptut an empty file? No file at all?

[af_trouble_orbits.txt](https://github.com/TerraFusion/advancedFusion/files/2082275/af_trouble_orbits.txt)

**Comment by LandonTClipp** [comment]
Oh and I should add, these changes are not necessary for me to continue progress. There was only a bug in the code that made it crash when it encountered missing datasets. I will leave it up to you how you want to use the macros.

**Comment by LandonTClipp** [comment]
6f021d11316d99b2c65beb8a62191dfa1e404f31
I have reverted changes and reimplemented the bug fix for this issue. I will close the ticket.

**Comment by jkimdevnc** [comment]
Hi Landon, Thanks a lot for the seg fault fix!   Also, thanks for taking care of the macro as well.  Good job.
